text/2020-01-30-key-spaces.md

```diff
@@ -0,0 +1,181 @@
+# Title
+
+Key Spaces
+
+## Summary
+
+When multiple applications store data in TiKV, we need a clear pattern for avoiding conflicting use of the key space in TiKV.
+TiKV is a monolithic KV store, so each application must choose a key prefix.
+Key Spaces formalizes this process by registering a key space for an application in TiKV itself.
+
+When combined with authentication (not part of this proposal), this can secure an application's key space.
+
+Additionally, key spaces make it possible for TiKV to avoid conflicts during a data restore by changing the key prefix assigned to a key space.
+
+## Motivation
+
+The use case for this is running multiple applications against TiKV (not just a single TiDB).
```

Comment:
I think this is not the recommended way

Reply:
Why?

Reply:
Is runing multiple applicactions against TiKV a strong demand?

Reply:
Its a chicken and egg problem. This is rarely done now because we lack basic features for it such as the one in the RFC. But users do ask about it.
For TiKV to be usable as a general purpose KV store these kinds of feature must be present. If it were usable as a general purpose KV store, users would start running KV workloads against it.

Reply:
I think there are two points to be discussed here.
* Why the feature is necessary?
People prefer simpler, newer tech-stack and fewer costs to achieve all goals, like using one database for all products. The HA and scalability make TiKV a good solution for some startups, however, they may do not have large amount of data, running multiple instances of TiDB on one TiKV cluster is a natural choice for them.
However, things are always hard, when multiple products rely on one TiKV cluster, data isolation is necessary for data security and resource usage. `Key Spaces` makes the data invisible to each other but it can not isolate the resource like CPU or memory usage. The latter may be solved by virtualization-tech.

* Why the feature should be put on TiKV?
As describes above. It is hard to isolate resource usage on TiKV and maybe the better solution is virtualization(VMs, Containers...). Another choice is to put this feature above TiKV which keeps TiKV simple and clean.

Reply:
Resource isolation is already an issue for TiDB users. TiDB itself supports multiple applications. And just storing data in 2 different tables can be a form of multi-tenant data with different workload requirements that impact each other. Even for a single table there can be transnational workloads and analytical workloads at the same time. So resource isolation already needs to be addressed and it is already up to users to determine how to handle this. After this RFC is implemented, it will still be up to users to determine how to achieve the workload isolation they need and balance that with cost and operational complexity. But it will be easier for them to experiment with different topologies.

Putting this feature client-side makes for less work for TiKV developers but nobody has explained what the advantage is otherwise. A client-side implementation will then fail to meet the requirements in the RFC.

Reply:
For the resource isolation in TiDB, 4.0 PD has a feature named placement rules, it supports balancing 2 different tables to different machines or different zone, this depends on the rule you specified.

Reply:
I updated the text here to better explain the need for key spaces. Please add your comments back if they still need to be addressed.